File: 88.合并两个有序数组/88.js
/**
 * @param {number[]} nums1
 * @param {number} m
 * @param {number[]} nums2
 * @param {number} n
 * @return {void} Do not return anything, modify nums1 in-place instead.
 */
// 超过：80.76%
var merge = function(nums1, m, nums2, n) {
    let size = m + n - 1
    while (n > 0) {
        // console.log(`init: `, {size, nums1: nums1[m - 1], nums2: nums2[n - 1]}, nums1, m, nums2, n)
        if (nums1[m - 1] > nums2[n - 1]) {
            nums1[size] = nums1[m - 1]
            m --
            size --
        } else if (m - 1 < 0 || nums1[m - 1] < nums2[n - 1]) {
            nums1[size] = nums2[n - 1]
            n --
            size --
        } else {
            nums1[size] = nums1[m - 1]
            m --
            size --
            nums1[size] = nums2[n - 1]
            n --
            size --
        }
        // console.log(`mid: `, {size, nums1: nums1[m - 1], nums2: nums2[n - 1]}, nums1, m, nums2, n)
    }
    return nums1
}

// console.log(merge([1, 2, 3, 0, 0, 0], 3, [2, 5, 6], 3))
// console.log(merge([1, 2, 3, 4, 6, 7], 6, [2, 5, 6], 3))
console.log(merge([4, 6, 8, 10, 13, 13], 6, [1, 2, 4, 7], 4))

File: 000001-000200/000088.合并两个有序数组/88.go
package main

import (
	"fmt"
)

// 方法一：用时 0ms
//// 执行用时超过了 100%
//// 消耗内存超过了
func merge(nums1 []int, m int, nums2 []int, n int)  {
	size := m + n - 1
	for i, j := m- 1, n - 1; i >= 0 || j >= 0; {
		if i == -1 {
			nums1[size] = nums2[j]
			size --
			j --
		} else if j == -1 {
			nums1[size] = nums1[i]
			size--
			i--
		} else if nums1[i] == nums2[j] {
			nums1[size] = nums1[i]
			size --
			i --
			nums1[size] = nums2[j]
			size --
			j --
		} else if nums1[i] > nums2[j] {
			nums1[size] = nums1[i]
			size --
			i --
		} else {
			nums1[size] = nums2[j]
			size --
			j --
		}
	}
	fmt.Println("num1:", nums1)
}

func main ()  {
	m := 3
	n := 6
	nums1 := []int{0, 0, 3, 0, 0, 0, 0, 0, 0}
	nums2 := []int{-1, 1, 1, 1, 2, 3}
	merge(nums1, m, nums2, n)
}
